FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs

```diff
@@ -0,0 +1,117 @@
+        public async Task<IActionResult> LikeNotificationAsync(string loggedInUser, string triggerUser, int postId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "like");
+        }
+
+        /// <summary>
+        /// API call from other services to create a follow notification. The user that was followed will get a notification.
+        /// </summary>
+        /// <param name="loggedInUser">email of user who has been followed (followee)</param>
+        /// <param name="triggerUser">email of the user who followed (follower)</param>
+        /// <param name="profileId">Profile ID of the trigger user</param>
+        /// <returns>Ok if accepted, BadRequest if fail</returns>
+        [HttpPost("follow")]
+        public async Task<IActionResult> FollowNotificationAsync(string loggedInUser, string triggerUser, int profileId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, profileId, "follow");
+        }
+
+        /// <summary>
+        /// Helper method to create the notification.
+        /// </summary>
+        /// <param name="loggedInUser">email of user that will receive the notification</param>
+        /// <param name="triggerUser">email of user that triggered the notification event</param>
+        /// <param name="linkId">Id of the resource to link to with the notification</param>
+        /// <param name="notificationType">The type of the notification (comment/follow/like)</param>
+        /// <returns></returns>
+        private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, string triggerUser, int linkId, string notificationType)
+        {
+            Notification notification = new Notification
+            {
+                Type = new System.Collections.Generic.KeyValuePair<string, int>(notificationType, linkId),
+                LoggedInUserId = loggedInUser,
+                TriggerUserId = triggerUser,
+                Date = DateTime.Now
+            };
+
+            string methodName = "";
+            if (notificationType == "comment")
+            {
+                methodName = "CommentNotification";
+            }
+            else if (notificationType == "like")
+            {
+                methodName = "LikeNotification";
+            }
+            else if (notificationType == "follow")
+            {
+                methodName = "FollowNotification";
+            }
+
+            try
+            {
+                await notificationsRepo.CreateNotificationAsync(notification);
+                User user = await userRepo.GetUserAsync(loggedInUser);
+
+                foreach (var connection in user.Connections)
+                {
+                    await hub.Groups.AddToGroupAsync(connection, user.Email);
+                }
+
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
+                return Ok();
+            }
+            catch (Exception e)
```

Comment:
not every exception is the fault of the client. unexpected exceptions are generally the server's fault.

this exception handling is actively bad in worse ways though: you're preventing the default logging of uncaught exceptions, so it's harder for the devs to detect an issue, while sending the exception to the client, so it's easy for hackers to detect an issue.